openmdao/core/problem.py

```diff
@@ -531,11 +567,170 @@ def run(self):
+        #             for d in all_vars[:-1]:
+        #                 myinputs.update(d)
+
+        if MPI:  # filter based on who owns the variables
+            # TODO Eventually, we think we can get rid of this next check. But to be safe,
+            #       we are leaving it in there.
+            if not model.is_active():
+                raise RuntimeError("RecordingManager.startup should never be called when "
+                                   "running in parallel on an inactive System")
+            rrank = self.comm.rank
+            rowned = model._owning_rank
+            mydesvars = [n for n in mydesvars if rrank == rowned[n]]
+            myobjectives = [n for n in myobjectives if rrank == rowned[n]]
+            myconstraints = [n for n in myconstraints if rrank == rowned[n]]
+
+        self._filtered_vars_to_record = {
+            'des': mydesvars,
+            'obj': myobjectives,
+            'con': myconstraints,
+        }
+
+        self._rec_mgr.startup(self)
+        if self.recording_options['record_metadata']:
+            self._rec_mgr.record_metadata(self)
+
+    def add_recorder(self, recorder):
+        """
+        Add a recorder to the problem.
+
+        Parameters
+        ----------
+        recorder : BaseRecorder
+           A recorder instance.
+        """
+        self._rec_mgr.append(recorder)
+
     def cleanup(self):
         """
         Clean up resources prior to exit.
         """
         self.driver.cleanup()
+        self._rec_mgr.close()
+
+    def record_iteration(self, case_name):
+        """
+        Record the variables at the Problem level.
+
+        Parameters
+        ----------
+        case_name : str
+            Name used to identify this Problem case.
+        """
+        if not self._rec_mgr._recorders:
+            return
+
+        # Get the data to record (collective calls that get across all ranks)
+        opts = self.recording_options
+        filt = self._filtered_vars_to_record
+
+        model = self.model
+        driver = self.driver
+
+        if opts['record_desvars']:
+            des_vars = driver.get_design_var_values()
+        else:
+            des_vars = {}
+
+        if opts['record_objectives']:
+            obj_vars = driver.get_objective_values()
+        else:
+            obj_vars = {}
+
+        if opts['record_constraints']:
+            con_vars = driver.get_constraint_values()
+        else:
+            con_vars = {}
+
+        des_vars = {name: des_vars[name] for name in filt['des']}
+        obj_vars = {name: obj_vars[name] for name in filt['obj']}
+        con_vars = {name: con_vars[name] for name in filt['con']}
+
+        names = model._outputs._names
+        views = model._outputs._views
+
+        if MPI:
+            des_vars = model._gather_vars(model, des_vars)
+            obj_vars = model._gather_vars(model, obj_vars)
+            con_vars = model._gather_vars(model, con_vars)
+
+        outs = {}
```

Comment:
Is there any reason why we couldn't combine des_vars, obj_vars, and con_vars into 'outs' BEFORE we call model._gather_vars, and just do a single MPI gather call using 'outs' rather than 3 different gather calls?  We do this same thing in Driver.record_iteration as well, except there we make 6 calls when it seems like we could get away with just 2.

Reply:
I put in a chore story to do this